plugins/auth-backend/src/providers/azure/provider.ts

```diff
@@ -0,0 +1,290 @@
+  }
+
+  async handler(
+    req: express.Request,
+  ): Promise<{ response: OAuthResponse; refreshToken: string }> {
+    const { response, privateInfo } = await executeFrameHandlerStrategy<
+      OAuthResponse,
+      PrivateInfo
+    >(req, this._strategy);
+    return {
+      response: await this.populateIdentity(response),
+      refreshToken: privateInfo.refreshToken,
+    };
+  }
+
+  async refresh(refreshToken: string, scope: string): Promise<OAuthResponse> {
+    const { accessToken, params } = await executeRefreshTokenStrategy(
+      this._strategy,
+      refreshToken,
+      scope,
+    );
+
+    const profile = await executeFetchUserProfileStrategy(
+      this._strategy,
+      accessToken,
+      params.id_token,
+    );
+
+    return this.populateIdentity({
+      providerInfo: {
+        accessToken,
+        idToken: params.id_token,
+        expiresInSeconds: params.expires_in,
+        scope: params.scope,
+      },
+      profile,
+    });
+  }
+
+  private async populateIdentity(
+    response: OAuthResponse,
+  ): Promise<OAuthResponse> {
+    const { profile } = response;
+    if (!profile.email) {
+      throw new Error('Azure AD profile contained no email.');
+    }
+
+    // As per Google implementation, set to local part of email for now
+    const id = profile.email.split('@')[0];
+
+    return { ...response, backstageIdentity: { id } };
+  }
+}
+
+export function createAzureProvider(
+  { baseUrl }: AuthProviderConfig,
+  _: string,
+  envConfig: Config,
+  logger: Logger,
+  tokenIssuer: TokenIssuer,
+) {
+  const providerId = 'azure';
+  const secure = envConfig.getBoolean('secure');
+  const appOrigin = envConfig.getString('appOrigin');
+
+  const clientID = envConfig.getString('clientId');
+  const clientSecret = envConfig.getString('clientSecret');
+  const tenantID = envConfig.getString('tenantId');
+
+  // URL metadata (e.g. authorize and token endpoints) are retrieved from the OIDC config endpoint
+  const identityMetadata = `https://login.microsoftonline.com/${tenantID}/v2.0/.well-known/openid-configuration`;
+
+  const responseType = 'code';
+  const responseMode = 'query';
+  const redirectUrl = `${baseUrl}/${providerId}/handler/frame`;
+
+  /**
+   * Space separated list of required scope values. 'openid' is sent by default so does not need to be specified here.
+   * For more info, see:
+   * - https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-permissions-and-consent
+   * - https://docs.microsoft.com/en-us/graph/permissions-reference
+   */
+  const scope = 'offline_access profile email User.Read';
+
+  const passReqToCallback = true;
+
+  // passport-azure-ad requires use of cookies to be session-free
+  const useCookieInsteadOfSession = true;
+  const cookieEncryptionKeys = [
+    { key: '12345678901234567890123456789012', iv: '123456789012' },
+    { key: 'abcdefghijklmnopqrstuvwxyzabcdef', iv: 'abcdefghijkl' },
```

Comment:
`passport-azure-ad` requires `express-session` or the use of cookies to be sessionless.